src/detect-engine-alert.c

```diff
@@ -188,44 +231,54 @@ int PacketAlertRemove(Packet *p, uint16_t pos)
 int PacketAlertAppend(DetectEngineThreadCtx *det_ctx, const Signature *s,
         Packet *p, uint64_t tx_id, uint8_t flags)
 {
-    int i = 0;
+    SCLogDebug("sid %" PRIu32 "", s->id);
+
+    /* highly unlikely, but seems better to check */
+    BUG_ON(p->alerts.cnt > packet_alert_max);
+
+    if (p->alerts.cnt < packet_alert_max) {
+        /* It should be usually the last, so check it before iterating */
+        /* Same signatures can generate more than one alert, if it's a diff tx */
+        if (p->alerts.cnt == 0 || p->alerts.alerts[p->alerts.cnt - 1].num <= s->num) {
+            /* We just add it */
+            PacketAlertInsertPos(det_ctx, s, p, tx_id, flags, p->alerts.cnt);
+            SCLogDebug("Added signature %" PRIu32 " to packet %" PRIu64 "", s->id, p->pcap_cnt);
+        } else {
+            /* find position to insert */
+            int16_t i = MemMovePacketAlertQueuePos(p, s);
+            PacketAlertInsertPos(det_ctx, s, p, tx_id, flags, i);
+            SCLogDebug("Added signature %" PRIu32 " to packet %" PRIu64 "", s->id, p->pcap_cnt);
+        }
 
-    if (p->alerts.cnt == PACKET_ALERT_MAX)
-        return 0;
-
-    SCLogDebug("sid %"PRIu32"", s->id);
-
-    /* It should be usually the last, so check it before iterating */
-    if (p->alerts.cnt == 0 || (p->alerts.cnt > 0 &&
-                               p->alerts.alerts[p->alerts.cnt - 1].num < s->num)) {
-        /* We just add it */
-        p->alerts.alerts[p->alerts.cnt].num = s->num;
-        p->alerts.alerts[p->alerts.cnt].action = s->action;
-        p->alerts.alerts[p->alerts.cnt].flags = flags;
-        p->alerts.alerts[p->alerts.cnt].s = s;
-        p->alerts.alerts[p->alerts.cnt].tx_id = tx_id;
-        p->alerts.alerts[p->alerts.cnt].frame_id =
-                (flags & PACKET_ALERT_FLAG_FRAME) ? det_ctx->frame_id : 0;
+        /* Update the count */
+        p->alerts.cnt++;
     } else {
-        /* We need to make room for this s->num
-         (a bit ugly with memcpy but we are planning changes here)*/
-        for (i = p->alerts.cnt - 1; i >= 0 && p->alerts.alerts[i].num > s->num; i--) {
-            memcpy(&p->alerts.alerts[i + 1], &p->alerts.alerts[i], sizeof(PacketAlert));
+        SCLogDebug("Reached packet_alert_max.");
+        /* If we reach packet_alert_max, remove lower priority
+         * rules and keep newer, higher priority ones.
+         * If Suri wants to append a signature whose priority is lower than the
+         * ones already queued and we are at packet_alert_max, it isn't queued. */
+        int16_t num_diff = s->num - p->alerts.alerts[p->alerts.cnt - 1].num;
+        if (num_diff == 0 || num_diff == -1) {
```

Comment:
Why have `num_diff == 0` ?

Reply:
because we can have the same signature appended twice to the same packet by different transactions.

Reply:
Ok thanks, but in that case, why do you want to keep the newest ?

Reply:
I assumed it would be better to have newer alerts.
It is, to be honest, unclear what is the best approach here. 
According to Victor, this part of the work will probably go to backports, and we would have something more flexible for 7. But at some point, for either approach, we may reach packet-alert-max, and then we have to decide on what to keep. 
We will have stats saying how many alerts were discarded, as well as do something about rules with the `drop` action, but... what to do when we have the same signature, triggered for the same packet, by different transactions? How to decide? Would it be good to have the perspective of threat hunters?